[BUGFIX] [FRONTEND] Correct chat logprobs

br3no opened the pull request:
FIX #5008 – correct `logprobs` format

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

br3no committed:
added correct models. Tests still missing

br3no committed:
fix implementation and tests

br3no committed:
fix formatting

br3no committed:
fix test

br3no commented:
Seems the failing builds are only due to some infrastructure issues...

DarkLight1337 commented:
I think `stop_reason` was intentionally added in #2976. Better not remove it from the output.

DarkLight1337 commented:
Also, the existing code for `_create_logprobs` fails the type checker (at least when I'm developing on VSCode). Feel free to use my updated implementation.

br3no commented:
> I think `stop_reason` was intentionally added in #2976. Better not remove it from the output.

Good catch. 

But this means that some clients might break. I have to test this, but I think clients in statically typed languages will not work with this, since it’s not part of the official API.

br3no commented:
> Also, the existing code for `_create_logprobs` fails the type checker (at least when I'm developing on VSCode). Feel free to use my updated implementation.

Can you share some details about what the type checker is complaining about? I'm not seeing any errors.

DarkLight1337 commented:
> > Also, the existing code for `_create_logprobs` fails the type checker (at least when I'm developing on VSCode). Feel free to use my updated implementation.
> 
> Can you share some details about what the type checker is complaining about? I'm not seeing any errors.

Currently, cross-directory checking is disabled (`--follow-imports=skip`) which suppresses most errors. @rkooo567 is working on properly type annotating the code before enabling cross-directory checking. To avoid introducing more errors, I've developing with cross-directory checking enabled. Specifically inside the context of `OpenAIServing._create_logprobs` and `OpenAIServingChat._get_top_logprobs`, I'm getting the following errors on your branch:

mypy
```
vllm/entrypoints/openai/serving_engine.py:110: error: Key expression in dictionary comprehension has incompatible type "Optional[str]"; expected type "str"  [misc]

vllm/entrypoints/openai/serving_chat.py:415: error: Argument "token" to "ChatCompletionLogProb" has incompatible type "Optional[str]"; expected "str"  [arg-type]
vllm/entrypoints/openai/serving_chat.py:417: error: Item "None" of "Optional[str]" has no attribute "encode"  [union-attr]
```

pyright
```
vllm/entrypoints/openai/serving_engine.py:102:40 - error: Argument of type "str | None" cannot be assigned to parameter "object" of type "str" in function "append"
  Type "str | None" cannot be assigned to type "str"
    "None" is incompatible with "str" (reportArgumentType)
vllm/entrypoints/openai/serving_engine.py:107:50 - error: Argument of type "dict[str | None, float] | None" cannot be assigned to parameter "object" of type "Dict[str, float] | None" in function "append" (reportArgumentType)
vllm/entrypoints/openai/serving_engine.py:119:34 - error: Argument of type "str | None" cannot be assigned to parameter "obj" of type "Sized" in function "len"
  Type "str | None" cannot be assigned to type "Sized"
    "None" is incompatible with protocol "Sized"
      "__len__" is not present (reportArgumentType)
vllm/entrypoints/openai/serving_engine.py:196:40 - error: Argument of type "str | None" cannot be assigned to parameter "text" of type "TextInput | PreTokenizedInput | List[PreTokenizedInput]" in function "__call__" (reportArgumentType)

vllm/entrypoints/openai/serving_chat.py:415:23 - error: Argument of type "str | None" cannot be assigned to parameter "token" of type "str" in function "__init__"
  Type "str | None" cannot be assigned to type "str"
    "None" is incompatible with "str" (reportArgumentType)
vllm/entrypoints/openai/serving_chat.py:417:44 - error: "encode" is not a known member of "None" (reportOptionalMemberAccess)
```

simon-mo commented:
@DarkLight1337 if you can help review and shepherd this PR that would be great (I'll merge upon your approval)

Regarding the output compliance if it is proven wide spread (can type safe languages to ignore extra fields? Otherwise it's difficult for forward compatibility I would have imagine), we can maybe return it via extra headers. (cc @njhill)

njhill commented:
Re additional fields in outputs, I feel that sticking to only what's supported by the current OpenAI APIs will be too restrictive. Just like how we have a bunch of extra request parameters (though I understand that doesn't have the same client compatibility concern). AFAIK clients can ignore unrecognized response fields in most cases.

If this turns out to be a problem, we could consider adding a "strict API response compatibility" config flag, and omit any extras if it's set?

br3no commented:
I have tested today with the Azure OpenAI client for Java (cf. https://platform.openai.com/docs/libraries/azure-openai-libraries) and it worked. So at least with this client the extra `stop_reason` is not a problem.

While this is not an exhaustive proof, I'll add it back. Then I'll see that I merge the changes from @DarkLight1337 into this PR.

br3no committed:
reverting removal of

br3no committed:
refactoring – move 'create_logprobs' for completion out of serving_engine.py

br3no committed:
fix formatting

br3no committed:
adding changes after review from @DarkLight1337

br3no commented:
@DarkLight1337 

I've added the following changes following your PR:
1. in `protocol.py` `top_logprobs` is no longer required  to be defined when `logprobs` is `True`
2. `LogProbs` has been renamed into `CompletionLogProbs`
3. improved tests
4. fixed code for type checker

There are some parts in your PR that I disagree with:
1. in `protocol.py` the format of the `logprobs` object for the chat completion is not in accordance to the OpenAI API spec. You can see in the official API spec [here](https://github.com/openai/openai-openapi/blob/893ba52242dbd5387a97b96444ee1c742cfce9bd/openapi.yaml#L460) that:
   1. the `logprobs` object contains a `content` array
   2. the `content` array contains objects with a `token` string, a `logprob` number, a `bytes` array of ints and a `top_logprobs` array of objects
   5. the `top_logprobs` array of objects contain a `token` string, a `logprob` number and a `bytes` array of ints
2. also, when requesting `"top_logprobs" : 2`, the `top_logprobs` array contains the top token (redundantly) plus the second best token. I.e. `len(top_logprobs) == k`. This can also be seen in the link above.

DarkLight1337 commented:
> There are some parts in your PR that I disagree with:

Thanks for going into depth regarding the API and raising these issues!

> 1. in `protocol.py` the format of the `logprobs` object for the chat completion is not in accordance to the OpenAI API spec. You can see in the official API spec [here](https://github.com/openai/openai-openapi/blob/893ba52242dbd5387a97b96444ee1c742cfce9bd/openapi.yaml#L460) that:
>    
>    1. the `logprobs` object contains a `content` array
>    2. the `content` array contains objects with a `token` string, a `logprob` number, a `bytes` array of ints and a `top_logprobs` array of objects
>    3. the `top_logprobs` array of objects contain a `token` string, a `logprob` number and a `bytes` array of ints

My PR does follow this, as specified in `vllm.entrypoints.openai.protocol.ChatCompletionLogProbs`. Perhaps you missed it?

> 2. also, when requesting `"top_logprobs" : 2`, the `top_logprobs` array contains the top token (redundantly) plus the second best token. I.e. `len(top_logprobs) == k`. This can also be seen in the link above.

I've looked through the docs again, and it indeed seems that `len(top_logprobs) == k` for Chat Completions API. I'll edit my message in #5008 accordingly.

DarkLight1337 commented:
> According to this, an object of type `vllm.entrypoints.openai.protocol.ChatCompletionLogProbs` contains, optionally, a list of `vllm.entrypoints.openai.protocol.ChatCompletionLogProb`. Each element in the list contains only one field called top_logprobs,

Notice that `ChatCompletionLogProb` inherits from `ChatCompletionTopLogProb`.

br3no committed:
Merge branch 'main' into 5008-chat-logprobs

br3no committed:
review iteration 2

br3no commented:
@DarkLight1337 I believe all issues are now addressed. Thanks for your help!

br3no committed:
formatting – isort breaks it again..?

br3no commented:
I'm having an issue with the formatting. It seems the formatter wants the imports in a different way than isort...

br3no commented:
isort wants this:
<img width="603" alt="image" src="https://github.com/vllm-project/vllm/assets/30877433/e3fdf4e1-80a9-44f1-bec0-76a7763ca4f6">

yapf wants this:
<img width="548" alt="image" src="https://github.com/vllm-project/vllm/assets/30877433/ddd393fa-ddb0-46e6-a4cb-0f2fba38e17f">

I don't really know how to fix this conflict... @DarkLight1337 can you help?

br3no commented:
One could set the multi-line mode in isort to https://pycqa.github.io/isort/docs/configuration/multi_line_output_modes.html#4-hanging-grid. But this would be a project-wide change. I'm uncomfortable doing it.

simon-mo commented:
I ran it locally and it seems the following format passed `./format.sh`

```
diff --git a/vllm/entrypoints/openai/serving_completion.py b/vllm/entrypoints/openai/serving_completion.py
index cfa2e8ae..7bc5a127 100644
--- a/vllm/entrypoints/openai/serving_completion.py
+++ b/vllm/entrypoints/openai/serving_completion.py
@@ -8,10 +8,13 @@ from fastapi import Request
 
 from vllm.config import ModelConfig
 from vllm.engine.async_llm_engine import AsyncLLMEngine
-from vllm.entrypoints.openai.protocol import (
-    CompletionLogProbs, CompletionRequest, CompletionResponse,
-    CompletionResponseChoice, CompletionResponseStreamChoice,
-    CompletionStreamResponse, UsageInfo)
+from vllm.entrypoints.openai.protocol import (CompletionLogProbs,
+                                              CompletionRequest,
+                                              CompletionResponse,
+                                              CompletionResponseChoice,
+                                              CompletionResponseStreamChoice,
+                                              CompletionStreamResponse,
+                                              UsageInfo)
```

DarkLight1337 commented:
Apart from the formatting issue, LGTM. @simon-mo do you have permission to update their fork? If so, you can apply your suggestion and then merge this request.

br3no commented:
@simon-mo this is the formatting produced by isort after the yapf formatting. 

When run locally, there is no check whether yapf or isort changed the sources. When run in the CI environment, the fact that yapf changes this leads to build errors.

simon-mo commented:
@rkooo567 any suggestion to this?

DarkLight1337 commented:
There are some other places where yapf is disabled via comments. Perhaps you have to do that here.

br3no committed:
disable yapf in import to avoid conflict with isort

DarkLight1337 commented:
While you're here, please also resolve the merge conflict.

br3no committed:
Merge branch 'main' into 5008-chat-logprobs

br3no committed:
fix formatting

simon-mo commented:
#5031 was just merged which seems to create conflict

br3no committed:
Merge branch 'main' into 5008-chat-logprobs

br3no committed:
fix merge error

br3no commented:
@simon-mo conflicts resolved. Build should be through soon.

br3no commented:
@simon-mo all tests green. We can merge!

ywang96 commented:
@br3no I'll merge this on behalf of Simon - thank you for fixing this issue!
